File: 4.0 Code/lidar_counter.py
from rplidar import RPLidar
from config import *
import time
from state import shared_state

def start_lidar():
    lidar = RPLidar(LIDAR_PORT)
    lidar.start_motor()
    print("Lidar running")

    entry_active = False
    exit_active = False

    try:
        for scan in lidar.iter_scans():
            for _, angle, distance in scan:

                if distance <= 0 or distance > TRIGGER_DISTANCE:
                    continue

                # ===== ENTRY LINE =====
                if abs(angle - ENTRY_LINE) < ANGLE_WINDOW:
                    if not entry_active:
                        with shared_state.lock:
                            if shared_state.car_count < MAX_SPOTS:
                                shared_state.car_count += 1
                                shared_state.last_event = "ENTER"
                                shared_state.last_event_time = time.strftime("%I:%M %p")
                        entry_active = True
                        print(f"ENTER → {shared_state.car_count}")
                else:
                    entry_active = False

                # ===== EXIT LINE =====
                if abs(angle - EXIT_LINE) < ANGLE_WINDOW:
                    if not exit_active:
                        with shared_state.lock:
                            if shared_state.car_count > 0:
                                shared_state.car_count -= 1
                                shared_state.last_event = "EXIT"
                                shared_state.last_event_time = time.strftime("%I:%M %p")
                        exit_active = True
                        print(f"EXIT → {shared_state.car_count}")
                else:
                    exit_active = False

    except Exception as e:
        print("Lidar error:", e)

    finally:
        lidar.stop_motor()
        lidar.stop()
        lidar.disconnect()

File: C++/Lidar_counter.cpp
// lidar.cpp
#include "config.h"
#include "state.h"
#include <iostream>
#include <chrono>
#include <thread>

// Placeholder — will connect to real SDK
float fake_read_angle() { return rand() % 360; }
float fake_read_distance() { return rand() % 4000; }

void start_lidar() {
    std::cout << "Lidar running" << std::endl;

    bool entry_active = false;
    bool exit_active = false;

    auto last_entry_event = std::chrono::steady_clock::now();
    auto last_exit_event  = std::chrono::steady_clock::now();

    while (true) {
        float angle = fake_read_angle();
        float distance = fake_read_distance();

        if (distance <= 0 || distance > Config::TRIGGER_DISTANCE)
            continue;

        auto now = std::chrono::steady_clock::now();

        // ===== ENTRY =====
        if (fabs(angle - Config::ENTRY_LINE) < Config::ANGLE_WINDOW) {
            if (!entry_active) {
                if (std::chrono::duration_cast<std::chrono::seconds>(now - last_entry_event).count()
                    >= Config::EVENT_COOLDOWN_SECONDS) {

                    std::lock_guard<std::mutex> lock(shared_state.mtx);

                    if (shared_state.car_count < Config::MAX_SPOTS) {
                        shared_state.car_count++;
                        shared_state.last_event = "ENTER";
                        shared_state.last_event_time = "NOW";
                    }

                    last_entry_event = now;
                    std::cout << "ENTER → " << shared_state.car_count << std::endl;
                }
                entry_active = true;
            }
        } else {
            entry_active = false;
        }

        // ===== EXIT =====
        if (fabs(angle - Config::EXIT_LINE) < Config::ANGLE_WINDOW) {
            if (!exit_active) {
                if (std::chrono::duration_cast<std::chrono::seconds>(now - last_exit_event).count()
                    >= Config::EVENT_COOLDOWN_SECONDS) {

                    std::lock_guard<std::mutex> lock(shared_state.mtx);

                    if (shared_state.car_count > 0) {
                        shared_state.car_count--;
                        shared_state.last_event = "EXIT";
                        shared_state.last_event_time = "NOW";
                    }

                    last_exit_event = now;
                    std::cout << "EXIT → " << shared_state.car_count << std::endl;
                }
                exit_active = true;
            }
        } else {
            exit_active = false;
        }

        std::this_thread::sleep_for(std::chrono::milliseconds(10));
    }
}
